base/utf16.jl

```diff
@@ -1,10 +1,82 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
+#=
+@doc """
+@brief      Error messages for Unicode / UTF support
+""" ->
+=#
+#=
+@enum(UTF_ERR,
+      SHORT = 1,
+      CONT,
+      LONG,
+      NOT_LEAD,
+      NOT_TRAIL,
+      NOT_SURROGATE,
+      MISSING_SURROGATE,
+      INVALID,
+      SURROGATE,
+      NULL_TERMINATE,
+)
+=#
+
+const UTF_ERR_SHORT = 1
+const UTF_ERR_CONT  = 2
+const UTF_ERR_LONG  = 3
+const UTF_ERR_NOT_LEAD = 4
+const UTF_ERR_NOT_TRAIL = 5
+const UTF_ERR_NOT_SURROGATE = 6
+const UTF_ERR_MISSING_SURROGATE = 7
+const UTF_ERR_INVALID = 8
+const UTF_ERR_SURROGATE = 9
+const UTF_ERR_NULL_TERMINATE = 10
+const UTF_ERR_MAX = 10
+
+type UnicodeError <: Exception
+    msg::AbstractString
+end
+
+#=
+@doc """
+@brief      Throws UnicodeError with information about the specific error, location, and character
+
+@param[in]  errCode::UTF_ERR
+@param[in]  errPos:: Integer
+@param[in]  errChar::Integer
+
+@throws never returns, always throws UnicodeError
+""" ->
+=#
+function utf_errfunc(errCode::Integer, errPos::Integer, errChar::Integer)
+    const errMsgs = [
+    "invalid UTF-8 sequence starting at index $errPos (0x$(hex(errChar)) missing one or more continuation bytes)",
+    "invalid UTF-8 sequence starting at index $errPos (0x$(hex(errChar)) is not a continuation byte)",
+    "invalid UTF-8 sequence, overlong encoding starting at index $errPos (0x$(hex(errChar)))",
+    "not a leading Unicode surrogate character at index $errPos (0x$(hex(errChar)))",
+    "not a trailing Unicode surrogate character at index $errPos (0x$(hex(errChar)))",
+    "not a valid Unicode surrogate character at index $errPos (0x$(hex(errChar))",
+    "missing trailing Unicode surrogate character after index $errPos (0x$(hex(errChar)))",
+    "invalid Unicode character starting at index $errPos (0x$(hex(errChar)) > 0x10ffff)",
+    "surrogate encoding not allowed in UTF-8 or UTF-32, at index $errPos (0x$(hex(errChar)))",
+    "UTF16String data must be NULL-terminated"
+    ]
+    if errCode < 1 || errCode > UTF_ERR_MAX
+        throw(UnicodeError("Bad error code: $errCode, Pos = $errPos, Char = $errChar"))
```

Comment:
OK, that's what I'd changed it to after your comment, good to have confirmation that that's the best... (there was also ErrorException... wasn't sure exactly when to use that one)